Move VAOS previousPage logic to state

LanceGundersen opened the pull request:
## Description
Currently we have next and previous functions for each entry in our form flow files (newAppointmentFlow and newExpressCareRequestFlow). We've had several bugs where the previous logic wasn't in sync with the next logic and going back through the form sent users to the incorrect page. This PR moves the previous page logic into redux state management under an array `previousPages` within the newAppointment or expressCare state.

## Testing done
- Verified the workflows are adding and removing from the array as the user navigates through the form.

## Screenshots
New Appointment Flow
![image](https://user-images.githubusercontent.com/8315447/92126527-48e16b80-edce-11ea-9445-b6549032f251.png)

Express Care Flow
![image](https://user-images.githubusercontent.com/8315447/92126347-10419200-edce-11ea-8eb7-39d66b2754b9.png)

## Acceptance criteria
- [ ] Can navigate forward and backward through the new appointment flow
- [ ] Can navigate forward and backward through the express care flow

## Definition of done
- [-] Events are logged appropriately
- [-] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

LanceGundersen committed:
Add previousPage logic to state

- Removes references to previous page from newAppointmentFlow.js
- Adjusts the logic of the routeToPageInFlow action
- Adds previousPages array to state

LanceGundersen committed:
Move expresscare previousPage logic to state

- Removes references to previous page from newExpressCareRequestFlow.js
- Adjusts the logic of the routeToPageInFlow action
- Adds previousPages array to state

LanceGundersen committed:
Fix new appointment actions test

LanceGundersen committed:
fix new appointment reducer test

LanceGundersen committed:
Fix express care info page test

jbalboni commented on src/applications/vaos/new-appointment/newAppointmentFlow.js:
```diff
@@ -320,13 +311,11 @@ export default {
   appointmentTime: {
     url: '/new-appointment/appointment-time',
     next: 'contactInfo',
-    previous: 'vaFacility',
```
I think there are some more previous functions in this file that could be removed

LanceGundersen commented on src/applications/vaos/express-care/redux/reducer.js:
```diff
@@ -109,6 +110,11 @@ export default function expressCareReducer(state = initialState, action) {
+          previousPages:
+            action.direction === 'next'
+              ? state.newRequest.previousPages.concat([action.pageKey])
+              : state.newRequest.previousPages.pop() &&
```
Ah ok, yeah good call.

LanceGundersen committed:
Remove previous functions from newAppointmentFlow

LanceGundersen committed:
Implement feedback

- Removes await from getState() call
- remove mutation of the previousPages array

LanceGundersen committed:
Remove previous newappointmentflow tests

LanceGundersen committed:
Add previous route test

jbalboni commented on src/applications/vaos/express-care/redux/reducer.js:
```diff
@@ -109,6 +110,10 @@ export default function expressCareReducer(state = initialState, action) {
+          previousPages:
+            action.direction === 'next'
+              ? state.newRequest.previousPages.concat([action.pageKey])
+              : state.newRequest.previousPages.slice(0, -1),
```
I was playing around with this on a review instance and I think I found a case where the array doesn't work. If you use the browser's back button to go back a page, you end up having to click on the "Back" button on the form twice to go back.

I think using the object approach we started with will solve that, rather than trying to do anything like listening for history changes.

LanceGundersen replied:
This makes sense that it'd break 👎

LanceGundersen replied:
i will get working on the object use case

LanceGundersen committed:
Refactor state object and logic

jbalboni commented on src/applications/vaos/express-care/redux/actions.js:
```diff
@@ -145,14 +145,35 @@ export function routeToPageInFlow(flow, history, current, action) {
+        nextPage = flow[nextAction];
+      } else {
+        const nextStateKey = await nextAction(getState(), dispatch);
+        if (
```
I think you'll need to do the previousPages logic whether the nextAction is a string or a function.

LanceGundersen replied:
ah good catch

LanceGundersen replied:
This is outside of the string or function check already

jbalboni commented on src/applications/vaos/express-care/redux/actions.js:
```diff
@@ -145,14 +145,35 @@ export function routeToPageInFlow(flow, history, current, action) {
+          });
+        }
+        Object.assign(state.expressCare.newRequest.previousPages, {
+          [current]: nextStateKey,
```
So, interestingly, I think the way you have the key and value here makes `previousPages` more of a `nextPages` object.

This isn't wrong, it just shifts where some of the code is written. Written this way, it makes handling setting `home` a little easier, but it forces you to do a `find` below.

If you flip these around and do `[nextStateKey]: current`, then you have to handle `home` differently, but line 171 below becomes `const previousPage = state.expressCare.newRequest.previousPages[current];`.

LanceGundersen replied:
Understood.

LanceGundersen replied:
Adjust object style

jbalboni commented on src/applications/vaos/express-care/redux/actions.js:
```diff
@@ -145,14 +145,35 @@ export function routeToPageInFlow(flow, history, current, action) {
+        if (
+          Object.keys(state.expressCare.newRequest.previousPages).length <= 0
+        ) {
+          Object.assign(state.expressCare.newRequest.previousPages, {
```
`Object.assign` mutates the previousPages object here, which breaks the Redux rules. The Redux-y way to do this is to pass the page keys you need into the `FORM_PAGE_CHANGE_COMPLETED` action and update the object in the reducer.

LanceGundersen replied:
Ah yeah the redux rules. OK I can move to the ruducer.

LanceGundersen replied:
Moved to reducer.

LanceGundersen committed:
Move logic to reducer to avoid breaking redux rules

jbalboni commented on src/applications/vaos/express-care/redux/reducer.js:
```diff
@@ -104,11 +105,22 @@ export default function expressCareReducer(state = initialState, action) {
     }
     case FORM_PAGE_CHANGE_COMPLETED: {
+      if (Object.keys(state.newRequest.previousPages).length <= 0) {
+        Object.assign(state.newRequest.previousPages, {
```
This is still mutating `previousPages`, but the logic looks good. This is the pattern we tend to use for these kind of updates:

```js
let previousPages = state.newRequest.previousPages;

if (!Object.keys(previousPages).length) {
  previousPages = {
    ...previousPages,
    [action.pageKey]: 'home',
  };
}

if (action.direction === 'next') {
  previousPages = {
     ...previousPages,
     [action.pageKeyNext]: action.pageKey,
  }
}
```

LanceGundersen replied:
Done

jbalboni reviewed (commented):
Pretty minor nit, looks good otherwise.

LanceGundersen committed:
Prevent mutating of state

narin commented on src/applications/vaos/express-care/redux/reducer.js:
```diff
@@ -104,11 +105,25 @@ export default function expressCareReducer(state = initialState, action) {
+          [action.pageKeyNext]: action.pageKey,
+        };
+      }
       return {
         ...state,
         newRequest: {
           ...state.newRequest,
           pageChangeInProgress: false,
+          previousPages: state.newRequest.previousPages,
         },
```
Looks like you're passing the unmodified `state.newRequest.previousPages` here.  You should instead pass the updated `previousPages`.  e.g.

```js
return {
  ...state,
  newRequest: {
    ...state.newRequest,
    pageChangeInProgress: false,
    previousPages,
  },
};
```

Just another suggestion, but on line 108, maybe rename the new `previousPages` to something like `updatedPreviousPages`, then return that in the updated state on line 126.  Might be a little clearer.

narin commented on src/applications/vaos/new-appointment/redux/reducer.js:
```diff
@@ -207,9 +208,23 @@ export default function formReducer(state = initialState, action) {
       return {
         ...state,
         pageChangeInProgress: false,
+        previousPages: state.previousPages,
```
Same thing here.  You should return `previousPages`, not `state.previousPages`.

LanceGundersen committed:
Implement fixes from review

LanceGundersen committed:
Fix state bugs

narin commented:
@LanceGundersen I was doing a final run through and I noticed that the "Cancel" button on the first page of the Express Care flow doesn't seem to be working.  It should redirect the user back to the VAOS homepage.

LanceGundersen committed:
Add home check to page change start

LanceGundersen committed:
Remove unused import

jbalboni reviewed (approved):
I have one question about a test change, but otherwise looks good to me.

Since this branch is pretty far behind master, you may want to pull in the latest and make sure the unit tests pass before merging.

LanceGundersen commented on src/applications/vaos/tests/new-express-care-request/info-page.unit.spec.jsx:
```diff
@@ -62,9 +62,6 @@ describe('VAOS integration: Express Care info page', () => {
       ),
     ).to.exist;
 
-    fireEvent.click(screen.getByText('Cancel'));
```
Will do

LanceGundersen replied:
I think that was due to the cancel not working before but should be fixed now. Checking 👀

LanceGundersen committed:
Add the code ommited due to failing test
